packages/repository/src/relations/relation.helpers.ts

```diff
@@ -132,3 +140,133 @@ function isInclusionAllowed<T extends Entity, Relations extends object = {}>(
   debug('isInclusionAllowed for %j (relation %s)? %s', include, allowed);
   return allowed;
 }
+
+/**
+ * Returns an array of instances from the target map. The order of arrays is based on
+ * the order of sourceIds
+ *
+ * @param sourceIds - One value or array of values (of the target key)
+ * @param targetMap - a map that matches sourceIds with instances
+ */
+export function flattenMapByKeys<T>(
+  sourceIds: unknown[],
+  targetMap: Map<unknown, T>,
+): (T | undefined)[] {
+  const result: (T | undefined)[] = new Array(sourceIds.length);
+  // mongodb: use string as key of targetMap, and convert sourceId to strings
+  // to make sure it gets the related instances.
+  sourceIds.forEach((id, index) => {
+    const key = normalizeKey(id);
+    const target = targetMap.get(key);
+    result[index] = target;
+  });
+
+  return result;
+}
+
+/**
+ * Returns a map which maps key values(ids) to instances. The instances can be
+ * grouped by different strategies.
+ *
+ * @param list - an array of instances
+ * @param keyName - key name of the source
+ * @param reducer - a strategy to reduce inputs to single item or array
+ */
+export function buildLookupMap<Key, InType, OutType = InType>(
+  list: InType[],
+  keyName: StringKeyOf<InType>,
+  reducer: (accumulator: OutType | undefined, current: InType) => OutType,
+): Map<Key, OutType> {
+  const lookup = new Map<Key, OutType>();
+  for (const entity of list) {
+    // get a correct key value
+    const key = getKeyValue(entity, keyName) as Key;
+    // these 3 steps are to set up the map, the map differs according to the reducer.
+    const original = lookup.get(key);
+    const reduced = reducer(original, entity);
+    lookup.set(key, reduced);
+  }
+  return lookup;
+}
+
+/**
+ * Returns value of a keyName. Aims to resolve ObjectId problem of Mongo.
+ *
+ * @param model - target model
+ * @param keyName - target key that gets the value from
+ */
+export function getKeyValue(model: AnyObject, keyName: string) {
+  return normalizeKey(model[keyName]);
+}
+
+/**
+ * Workaround for MongoDB, where the connector returns ObjectID
+ * values even for properties configured with "type: string".
+ *
+ * @param rawKey
+ */
+export function normalizeKey(rawKey: unknown) {
+  if (isBsonType(rawKey)) {
+    return rawKey.toString();
+  }
+  return rawKey;
+}
+
+/**
+ * Returns an array of instances. For HasMany relation usage.
+ *
+ * @param acc
+ * @param it
+ */
+export function reduceAsArray<T>(acc: T[] | undefined, it: T) {
+  if (acc) acc.push(it);
+  else acc = [it];
+  return acc;
+}
+/**
+ * Returns a single of an instance. For HasOne and BelongsTo relation usage.
+ *
+ * @param _acc
+ * @param it
+ */
+export function reduceAsSingleItem<T>(_acc: T | undefined, it: T) {
+  return it;
+}
+
+/**
+ * Dedupe an array
+ * @param {Array} input an array
```

Comment:
I think you have to put a `-` after the param's name e.g. `@param {Array} input - an array`. Also `an array` of what?